### PR TITLE
fix: cannot create an insight onto a dashboard when no permissions

### PR DESCRIPTION
**Pull request by pauldambra**
## Problem

see this community slack thread: https://posthogusers.slack.com/archives/C01GLBKHKQT/p1671547613708909

When duplicating an insight we POST the insight to the insight API. One of the results is a tile is created to link the new insight to any dashboard the existing insight was on.

In the thread we discuss that may not be desired... but also that you can duplicate an insight onto a dashboard you don't have permissions to change. It turns out permissions were enforced when PATCHing an insight but not when POSTing

## Changes

* checks dashboard permissions on insight create
* checks dashboard permissions on dashboard duplication
* shows a toast when duplication fails

Screenshot 2022-12-20 at 16 49 26

## How did you test this code?

running it locally

### CODE REVIEW

File: posthog/api/dashboards/dashboard.py [hunk 1]
```diff
@@ -130,6 +130,20 @@ def create(self, validated_data: Dict, *args: Any, **kwargs: Any) -> Dashboard:
...
+            try:
+                existing_dashboard: Dashboard = Dashboard.objects.get(id=use_dashboard, team=team)
+            except Dashboard.DoesNotExist:
+                raise serializers.ValidationError({"use_dashboard": "Invalid value provided"})
+
+            if (
+                existing_dashboard.get_effective_privilege_level(self.context["request"].user.id)
+                == Dashboard.PrivilegeLevel.CAN_VIEW
+            ):
+                raise PermissionDenied(f"You don't have permission to add insights to dashboard: {use_dashboard}")
```
Comment (Twixes) [inline posthog/api/dashboards/dashboard.py:145]: To be in line with `insight.py`:
```suggestion
                raise PermissionDenied(f"You don't have permission to add insights to dashboard: {existing_dashboard.name}")
```

File: posthog/api/insight.py [hunk 2]
```diff
@@ -223,19 +223,27 @@ def create(self, validated_data: Dict, *args: Any, **kwargs: Any) -> Insight:
...
-        if dashboards is not None:
-            for dashboard in Dashboard.objects.filter(id__in=[d.id for d in dashboards]).all():
-                if dashboard.team != insight.team:
-                    raise serializers.ValidationError("Dashboard not found")
-
-                DashboardTile.objects.create(insight=insight, dashboard=dashboard, last_refresh=now())
-                insight.last_refresh = now()  # set last refresh if the insight is on at least one dashboard
+        for dashboard in dashboards_to_add:
+            DashboardTile.objects.create(insight=insight, dashboard=dashboard, last_refresh=now())
+            insight.last_refresh = now()  # set last refresh if the insight is on at least one dashboard
```
Comment (Twixes) [inline posthog/api/insight.py:246]: This is only a refactor, but I wonder, do we actually refresh the data when this happens?

File: posthog/api/test/dashboards/test_dashboard.py [hunk 3]
```diff
@@ -21,6 +21,13 @@
 class TestDashboard(APIBaseTest, QueryMatchingTest):
     def setUp(self) -> None:
         super().setUp()
+
+        self.organization.available_features = [
+            AvailableFeature.DASHBOARD_COLLABORATION,
+            AvailableFeature.DASHBOARD_PERMISSIONING,
+        ]
```
Comment (Twixes) [inline posthog/api/test/dashboards/test_dashboard.py:28]: Considering `test_cannot_duplicate_dashboard_with_no_permissions_to_edit_it` requires these, it should probably be in `ee/api/test/test_dashboard.py` (there all the features are already available)

**Review by Twixes — CHANGES_REQUESTED** [review]
Definitely deserves to be fixed! Overall this makes sense, I left some notes on the details though

**Comment by pauldambra** [comment]
going to queue this behind #13492

**Comment by Twixes** [comment]
Has this been queued somewhere else? @pauldambra 👀

**Pull request closed without merging**